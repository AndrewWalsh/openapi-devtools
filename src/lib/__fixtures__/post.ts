const post: chrome.devtools.network.Request = {
  getContent() {},
  cache: {},
  _resourceType: "xhr",
  request: {
    method: "POST",
    url: "https://www.example.com/v1/track",
    httpVersion: "http/2.0",
    headers: [
      {
        name: ":method",
        value: "POST",
      },
      {
        name: ":path",
        value: "/v1/track",
      },
      {
        name: ":scheme",
        value: "https",
      },
      {
        name: "accept",
        value: "*/*",
      },
      {
        name: "accept-encoding",
        value: "gzip, deflate, br",
      },
      {
        name: "accept-language",
        value: "en-GB,en-US;q=0.9,en;q=0.8",
      },
      {
        name: "anonymousid",
        value: "aa",
      },
      {
        name: "content-length",
        value: "6798",
      },
      {
        name: "content-type",
        value: "application/json",
      },
    ],
    queryString: [
      {
        name: "alt",
        value: "json",
      },
      {
        name: "key",
        value: "abc",
      },
    ],
    cookies: [
      {
        name: "VISITOR_INFO1_LIVE",
        value: "abc",
        path: "/",
        domain: ".example.com",
        expires: "2024-04-12T07:45:16.478Z",
        httpOnly: true,
        secure: true,
      },
      {
        name: "YSC",
        value: "abc",
        path: "/",
        domain: ".example.com",
        expires: "1969-12-31T23:59:59.000Z",
        httpOnly: true,
        secure: true,
      },
    ],
    headersSize: -1,
    bodySize: 6798,
    postData: {
      mimeType: "application/json",
      text: '{ "test": "integer" }'
    },
  },
  response: {
    status: 200,
    statusText: "OK",
    httpVersion: "http/2.0",
    headers: [
      {
        name: "access-control-expose-headers",
        value: "",
      },
      {
        name: "content-length",
        value: "2",
      },
      {
        name: "content-type",
        value: "text/plain; charset=utf-8",
      },
      {
        name: "date",
        value: "Sun, 15 Oct 2023 07:42:54 GMT",
      },
      {
        name: "server",
        value: "openresty/1.21.4.2",
      },
      {
        name: "vary",
        value: "Origin",
      },
    ],
    cookies: [],
    content: {
      size: 28,
      mimeType: "application/json",
    },
    redirectURL: "",
    headersSize: -1,
    bodySize: -1,
    _transferSize: 166,
  },
  serverIPAddress: "44.219.182.139",
  startedDateTime: "2023-10-15T07:42:54.694Z",
  time: 318.07399999186396,
  timings: {
    blocked: 1.4950000001639128,
    dns: 0.010000000000000009,
    ssl: 105.26800000000001,
    connect: 203.84,
    send: 0.21299999999999386,
    wait: 112.1749999562949,
    receive: 0.341000035405159,
  },
};

export const postHost = 'www.example.com';

export default post;
