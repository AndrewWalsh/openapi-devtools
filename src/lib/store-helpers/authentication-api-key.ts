// import { Authentication, AuthType } from "../../utils/types";
// import { getAuthType } from "./authentication";

// /**
//  * Helpers and utilities for endpoint Authentication details
//  * Modelled on a Security Scheme Object https://spec.openapis.org/oas/v3.1.0#security-scheme-object
//  */
// export interface APIKeyAuth extends Authentication {
//   type: "apiKey";
//   name: string;
// }

// export interface APIKeyAuthHeader extends APIKeyAuth {
//   in: "header";
// }

// export type APIKeyAuthHeaders = APIKeyAuthHeader | undefined;

// export const parseAPIKeyAuthHeader = (
//   header: chrome.devtools.network.Request["request"]["headers"][0]
// ): APIKeyAuthHeaders => {
//   const authType = getAuthType(header.value);
//   if (authType === "basic") {
//     const basicAuth: BasicAuthHeader = {
//       id: AuthType.BASIC,
//       type: "http",
//       in: "header",
//       scheme: "Basic",
//       description: "",
//     };
//     return basicAuth;
//   } else if (authType === "bearer") {
//     const bearerAuth: BearerAuthHeader = {
//       id: AuthType.BEARER,
//       type: "http",
//       in: "header",
//       scheme: "Bearer",
//       description: "",
//     };
//     return bearerAuth;
//   } else if (authType === "digest") {
//     const digestAuth: DigestAuthHeader = {
//       id: AuthType.DIGEST,
//       type: "http",
//       in: "header",
//       scheme: "Digest",
//       description: "",
//     };
//     return digestAuth;
//   }
// };
