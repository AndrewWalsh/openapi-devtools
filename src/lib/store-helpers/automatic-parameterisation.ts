import validator from "validator";

const validationRules = [validator.isUUID];

/**
 * Fast and low-effort rules-based parameterisation of pathnames alone
 * Designed to accompany a more computationally intensive analysis
 * Returns an array of indices of path parameters in a given pathname that matches a validation rule
 */
export const fastPathParameterIndices = (pathname: string): Array<number> => {
  const indices = [];
  const parts = pathname.split("/").slice(1);
  for (let idx = 0; idx < parts.length; idx++) {
    const part = parts[idx];
    if (validationRules.some((rule) => rule(part))) {
      indices.push(idx);
    }
  }
  return indices;
};
