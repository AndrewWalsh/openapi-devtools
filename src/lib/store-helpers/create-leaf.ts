import decodeUriComponent from "decode-uri-component";
import {
  createSchemaElseUndefined,
  entriesToJSONType,
  parseJSON,
} from "../../utils/helpers";
import { JSONType, Leaf } from "../../utils/types";
import determineAuthFromHAR from "./authentication";
import { filterIgnoreHeaders } from "../../utils/headers";
import type { Options } from "../RequestStore";
import type { Entry } from "har-format";

export type Params = {
  harRequest: Entry;
  responseBody: JSONType;
  options: Options;
};

function createLeaf({ harRequest, responseBody, options }: Params): Leaf {
  const { enableMoreInfo } = options;
  const authentication = determineAuthFromHAR(harRequest);
  harRequest.request.headers = filterIgnoreHeaders(harRequest.request.headers);
  harRequest.response.headers = filterIgnoreHeaders(
    harRequest.response.headers
  );
  const method = harRequest.request.method;
  const statusCode = harRequest.response.status.toString();
  const requestMime = harRequest.request.postData?.mimeType;
  const responseMime = harRequest.response.content.mimeType;
  const requestBody = parseJSON(harRequest.request.postData?.text);
  const requestHeaders = entriesToJSONType(harRequest.request.headers);
  const responseHeaders = entriesToJSONType(harRequest.response.headers);
  const queryParameters = entriesToJSONType(harRequest.request.queryString);
  const pathname = decodeUriComponent(new URL(harRequest.request.url).pathname);
  const leafPart: Leaf = {
    ...(authentication && { authentication }),
    pathname,
    methods: {
      [method]: {
        [statusCode]: {
          ...(requestMime && {
            request: {
              [requestMime]: {
                body: createSchemaElseUndefined(requestBody),
                ...(enableMoreInfo && { mostRecent: requestBody }),
              },
            },
          }),
          requestHeaders: createSchemaElseUndefined(requestHeaders),
          response: {
            [responseMime]: {
              body: createSchemaElseUndefined(responseBody),
              ...(enableMoreInfo && { mostRecent: responseBody }),
            },
          },
          responseHeaders: createSchemaElseUndefined(responseHeaders),
          queryParameters: createSchemaElseUndefined(queryParameters),
        },
      },
    },
  };
  return leafPart;
}

export default createLeaf;
