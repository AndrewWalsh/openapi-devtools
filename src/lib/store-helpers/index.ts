export { default as createLeaf } from "./create-leaf";
export { default as parameterise } from "./parameterise";
export { default as leafMapToRouterMap } from "./leafmap-to-routermap";
export { default as upsert } from "./upsert";
export { default as findPathnamesInRouter } from "./find-pathnames-in-router";
export { default as pruneRouter } from "./prune-router";
export { default as persistOptions } from "./persist-options";
export { default as determineAuthFromHAR } from "./authentication";
export { default as remove } from "./remove";
export * from "./merge";
export * from "./helpers";
