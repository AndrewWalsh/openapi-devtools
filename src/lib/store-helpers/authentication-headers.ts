import type { Entry } from 'har-format';
import { Authentication, AuthType } from "../../utils/types";
import { isAuthHeader } from "../../utils/headers";
import { parseHTTPAuthHeader, HTTP } from "./authentication-http";
import cookie from "cookie";

const AUTHORIZATION = "AUTHORIZATION";
const COOKIE = "COOKIE";

/**
 * Helpers and utilities for endpoint Authentication details
 * Modelled on a Security Scheme Object https://spec.openapis.org/oas/v3.1.0#security-scheme-object
 */
export interface APIKeyAuth extends Authentication {
  type: "apiKey";
  name: string;
}

export interface APIKeyAuthHeader extends APIKeyAuth {
  in: "header";
}

export interface APIKeyAuthCookie extends APIKeyAuth {
  in: "cookie";
}

export type APIKey = APIKeyAuthHeader | APIKeyAuthCookie;

// Include cookie keys as api keys when they contain a string below
const cookieSecurityNames = ["token", "session", "id", "jwt", "auth", "cookie", "key", "api"];

const parseAPIKeyAuthCookie = (cookieStr: string): Array<APIKeyAuthCookie> => {
  const cookies = cookie.parse(cookieStr);
  const filteredSecurityNames = Object.keys(cookies).filter((cookieKey) => cookieSecurityNames.some(str => cookieKey.toLowerCase().includes(str)));
  return filteredSecurityNames.map((cookieKey) => {
    const lowerName = cookieKey.toLowerCase();
    const upperName = cookieKey.toUpperCase();
    return {
      authType: AuthType.APIKEY_COOKIE_ + upperName,
      name: lowerName,
      type: "apiKey",
      in: "cookie",
      description: "",
    };
  });
};

type OpenAPIAuth = APIKey | HTTP;

/**
 * Extract all auth info from request headers
 */
export const parseAuthHeaders = (
  headers: Entry["request"]["headers"]
): Array<OpenAPIAuth> => {
  const visited = new Set<string>();
  let results: Array<OpenAPIAuth> = [];
  for (const header of headers) {
    const upperName = header.name.toUpperCase();
    const isAPIKey = isAuthHeader(upperName);
    const wasVisited = visited.has(upperName);
    // Type: http, in: header
    if (upperName === AUTHORIZATION && !wasVisited) {
      const httpAuth = parseHTTPAuthHeader(header.value);
      if (httpAuth) results.push(httpAuth);
    }
    // Type: apiKey, in: cookie
    if (upperName === COOKIE && !wasVisited) {
      results = results.concat(parseAPIKeyAuthCookie(header.value));
    }
    // Type: apiKey, in: header
    if (isAPIKey && !wasVisited) {
      visited.add(upperName);
      const apiKey: APIKeyAuthHeader = {
        authType: AuthType.APIKEY_HEADER_ + upperName,
        name: upperName,
        type: "apiKey",
        in: "header",
        description: "",
      };
      results.push(apiKey);
    }
  }
  return results;
};
