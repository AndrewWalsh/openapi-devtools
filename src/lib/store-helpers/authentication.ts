import { isEmpty } from "lodash";
import { Leaf } from "../../utils/types";
import { parseAuthHeaders } from "./authentication-headers";
import type { Entry } from 'har-format';

export const getAuthType = (auth: string) => {
  const split = auth.split(" ");
  if (!split.length) return "";
  return split[0].toLowerCase();
};

type DetermineAuthFromHAR = (
  harRequest: Entry
) => Leaf["authentication"] | undefined;
const determineAuthFromHAR: DetermineAuthFromHAR = (harRequest) => {
  const finalAuth: Leaf["authentication"] = {};
  const authItems = parseAuthHeaders(harRequest.request.headers);
  authItems.forEach((auth) => {
    finalAuth[auth.authType] = auth;
  });
  if (isEmpty(finalAuth)) return;
  return finalAuth;
};

export default determineAuthFromHAR;
