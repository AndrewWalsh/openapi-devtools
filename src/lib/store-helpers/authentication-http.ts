import { Authentication, AuthType } from "../../utils/types";
import { getAuthType } from "./authentication";

/**
 * Helpers and utilities for endpoint Authentication details
 * Modelled on a Security Scheme Object https://spec.openapis.org/oas/v3.1.0#security-scheme-object
 */
export interface HTTPAuth extends Authentication {
  type: "http";
}

export interface BearerAuthHeader extends HTTPAuth {
  scheme: "Bearer";
  in: "header";
}

export interface BasicAuthHeader extends HTTPAuth {
  scheme: "Basic";
  in: "header";
}

export interface DigestAuthHeader extends HTTPAuth {
  scheme: "Digest";
  in: "header";
}

export type HTTP =
  | DigestAuthHeader
  | BasicAuthHeader
  | BearerAuthHeader;

export const parseHTTPAuthHeader = (value: string): HTTP | undefined => {
  const authType = getAuthType(value);
  if (authType === "basic") {
    const basicAuth: BasicAuthHeader = {
      authType: AuthType.HTTP_HEADER_BASIC,
      type: "http",
      in: "header",
      scheme: "Basic",
      description: "",
    };
    return basicAuth;
  } else if (authType === "bearer") {
    const bearerAuth: BearerAuthHeader = {
      authType: AuthType.HTTP_HEADER_BEARER,
      type: "http",
      in: "header",
      scheme: "Bearer",
      description: "",
    };
    return bearerAuth;
  } else if (authType === "digest") {
    const digestAuth: DigestAuthHeader = {
      authType: AuthType.HTTP_HEADER_DIGEST,
      type: "http",
      in: "header",
      scheme: "Digest",
      description: "",
    };
    return digestAuth;
  }
};
