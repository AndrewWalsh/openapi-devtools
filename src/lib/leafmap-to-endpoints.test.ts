import { it, expect } from "vitest";
import leafMapToEndpoints from "./leafmap-to-endpoints";
import post, { postHost } from "./__fixtures__/post";
import { createLeaf } from "./store-helpers";
import { LeafMap } from "../utils/types";

it("flattens data into endpoints", () => {
  const leafMap: LeafMap = {
    [postHost]: {
      "/1/2/3": { data: createLeaf(post, { foo: "bar" }) },
      "/1/:param1/3": { data: createLeaf(post, { foo: "bar" }) },
      "/1/2/3/4": { data: createLeaf(post, { foo: "bar" }) },
      "/1/2/3/4a": { data: createLeaf(post, { foo: "bar" }) },
      "/1/2/3/": { data: createLeaf(post, { foo: "bar" }) },
      "/1/2/3/:param4/param5/a/:param7/b": { data: createLeaf(post, { foo: "bar" }) },
    },
  };
  const endpoints = leafMapToEndpoints(leafMap);
  expect(endpoints).toHaveLength(6);
  expect(endpoints).toMatchSnapshot();
});
