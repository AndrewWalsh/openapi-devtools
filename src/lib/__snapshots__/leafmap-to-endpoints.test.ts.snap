// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flattens data into endpoints 1`] = `
[
  {
    "data": {
      "methods": {
        "POST": {
          "200": {
            "queryParameters": {
              "properties": {
                "alt": {
                  "type": "string",
                },
                "key": {
                  "type": "string",
                },
              },
              "required": [
                "alt",
                "key",
              ],
              "type": "object",
            },
            "request": {
              "application/json": {
                "body": {
                  "properties": {
                    "test": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "test",
                  ],
                  "type": "object",
                },
              },
            },
            "requestHeaders": {
              "properties": {
                "anonymousid": {
                  "type": "string",
                },
              },
              "required": [
                "anonymousid",
              ],
              "type": "object",
            },
            "response": {
              "application/json": {
                "body": {
                  "properties": {
                    "foo": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "foo",
                  ],
                  "type": "object",
                },
              },
            },
            "responseHeaders": undefined,
          },
        },
      },
      "pathname": "/v1/track",
    },
    "host": "www.example.com",
    "parts": [
      {
        "part": "1",
        "type": 0,
      },
      {
        "part": "2",
        "type": 0,
      },
      {
        "part": "3",
        "type": 0,
      },
    ],
    "pathname": "/1/2/3",
  },
  {
    "data": {
      "methods": {
        "POST": {
          "200": {
            "queryParameters": {
              "properties": {
                "alt": {
                  "type": "string",
                },
                "key": {
                  "type": "string",
                },
              },
              "required": [
                "alt",
                "key",
              ],
              "type": "object",
            },
            "request": {
              "application/json": {
                "body": {
                  "properties": {
                    "test": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "test",
                  ],
                  "type": "object",
                },
              },
            },
            "requestHeaders": {
              "properties": {
                "anonymousid": {
                  "type": "string",
                },
              },
              "required": [
                "anonymousid",
              ],
              "type": "object",
            },
            "response": {
              "application/json": {
                "body": {
                  "properties": {
                    "foo": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "foo",
                  ],
                  "type": "object",
                },
              },
            },
            "responseHeaders": undefined,
          },
        },
      },
      "pathname": "/v1/track",
    },
    "host": "www.example.com",
    "parts": [
      {
        "part": "1",
        "type": 0,
      },
      {
        "part": ":param1",
        "type": 1,
      },
      {
        "part": "3",
        "type": 0,
      },
    ],
    "pathname": "/1/:param1/3",
  },
  {
    "data": {
      "methods": {
        "POST": {
          "200": {
            "queryParameters": {
              "properties": {
                "alt": {
                  "type": "string",
                },
                "key": {
                  "type": "string",
                },
              },
              "required": [
                "alt",
                "key",
              ],
              "type": "object",
            },
            "request": {
              "application/json": {
                "body": {
                  "properties": {
                    "test": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "test",
                  ],
                  "type": "object",
                },
              },
            },
            "requestHeaders": {
              "properties": {
                "anonymousid": {
                  "type": "string",
                },
              },
              "required": [
                "anonymousid",
              ],
              "type": "object",
            },
            "response": {
              "application/json": {
                "body": {
                  "properties": {
                    "foo": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "foo",
                  ],
                  "type": "object",
                },
              },
            },
            "responseHeaders": undefined,
          },
        },
      },
      "pathname": "/v1/track",
    },
    "host": "www.example.com",
    "parts": [
      {
        "part": "1",
        "type": 0,
      },
      {
        "part": "2",
        "type": 0,
      },
      {
        "part": "3",
        "type": 0,
      },
      {
        "part": "4",
        "type": 0,
      },
    ],
    "pathname": "/1/2/3/4",
  },
  {
    "data": {
      "methods": {
        "POST": {
          "200": {
            "queryParameters": {
              "properties": {
                "alt": {
                  "type": "string",
                },
                "key": {
                  "type": "string",
                },
              },
              "required": [
                "alt",
                "key",
              ],
              "type": "object",
            },
            "request": {
              "application/json": {
                "body": {
                  "properties": {
                    "test": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "test",
                  ],
                  "type": "object",
                },
              },
            },
            "requestHeaders": {
              "properties": {
                "anonymousid": {
                  "type": "string",
                },
              },
              "required": [
                "anonymousid",
              ],
              "type": "object",
            },
            "response": {
              "application/json": {
                "body": {
                  "properties": {
                    "foo": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "foo",
                  ],
                  "type": "object",
                },
              },
            },
            "responseHeaders": undefined,
          },
        },
      },
      "pathname": "/v1/track",
    },
    "host": "www.example.com",
    "parts": [
      {
        "part": "1",
        "type": 0,
      },
      {
        "part": "2",
        "type": 0,
      },
      {
        "part": "3",
        "type": 0,
      },
      {
        "part": "4a",
        "type": 0,
      },
    ],
    "pathname": "/1/2/3/4a",
  },
  {
    "data": {
      "methods": {
        "POST": {
          "200": {
            "queryParameters": {
              "properties": {
                "alt": {
                  "type": "string",
                },
                "key": {
                  "type": "string",
                },
              },
              "required": [
                "alt",
                "key",
              ],
              "type": "object",
            },
            "request": {
              "application/json": {
                "body": {
                  "properties": {
                    "test": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "test",
                  ],
                  "type": "object",
                },
              },
            },
            "requestHeaders": {
              "properties": {
                "anonymousid": {
                  "type": "string",
                },
              },
              "required": [
                "anonymousid",
              ],
              "type": "object",
            },
            "response": {
              "application/json": {
                "body": {
                  "properties": {
                    "foo": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "foo",
                  ],
                  "type": "object",
                },
              },
            },
            "responseHeaders": undefined,
          },
        },
      },
      "pathname": "/v1/track",
    },
    "host": "www.example.com",
    "parts": [
      {
        "part": "1",
        "type": 0,
      },
      {
        "part": "2",
        "type": 0,
      },
      {
        "part": "3",
        "type": 0,
      },
      {
        "part": "",
        "type": 0,
      },
    ],
    "pathname": "/1/2/3/",
  },
  {
    "data": {
      "methods": {
        "POST": {
          "200": {
            "queryParameters": {
              "properties": {
                "alt": {
                  "type": "string",
                },
                "key": {
                  "type": "string",
                },
              },
              "required": [
                "alt",
                "key",
              ],
              "type": "object",
            },
            "request": {
              "application/json": {
                "body": {
                  "properties": {
                    "test": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "test",
                  ],
                  "type": "object",
                },
              },
            },
            "requestHeaders": {
              "properties": {
                "anonymousid": {
                  "type": "string",
                },
              },
              "required": [
                "anonymousid",
              ],
              "type": "object",
            },
            "response": {
              "application/json": {
                "body": {
                  "properties": {
                    "foo": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "foo",
                  ],
                  "type": "object",
                },
              },
            },
            "responseHeaders": undefined,
          },
        },
      },
      "pathname": "/v1/track",
    },
    "host": "www.example.com",
    "parts": [
      {
        "part": "1",
        "type": 0,
      },
      {
        "part": "2",
        "type": 0,
      },
      {
        "part": "3",
        "type": 0,
      },
      {
        "part": ":param4",
        "type": 1,
      },
      {
        "part": "param5",
        "type": 0,
      },
      {
        "part": "a",
        "type": 0,
      },
      {
        "part": ":param7",
        "type": 1,
      },
      {
        "part": "b",
        "type": 0,
      },
    ],
    "pathname": "/1/2/3/:param4/param5/a/:param7/b",
  },
]
`;
